File: packages/var-core/src/step-role.ts
// The role a step definition plays, mirroring concepts/sensors-and-actuators.md:
//   context — the quiescent state the software rests in
//   action  — the actuator: the single stimulus
//   sensor  — the read-only assertion (the only role that returns for comparison)
export type StepKind = 'context' | 'action' | 'sensor'

// Guess a step's role from its neighbours, using the canonical document order
// context → action → sensor. Purely structural — never inspects sentence words
// (no Given/When/Then heuristics). The generated snippet always offers the other
// roles as commented alternatives, so a wrong guess is cheap to correct.
export function inferStepRole(neighbours: {
  readonly before: ReadonlyArray<StepKind>
  readonly after: ReadonlyArray<StepKind>
}): StepKind {
  const { before, after } = neighbours
  if (after.length === 0) return 'sensor'
  if (after.includes('sensor') && !before.includes('action') && !after.includes('action')) {
    return 'action'
  }
  if (before.length === 0) return 'context'
  return 'action'
}

File: java/var-core/src/main/java/com/oselvar/var/core/StepRole.java
package com.oselvar.var.core;

import java.util.List;

/**
 * Port of {@code var-core/src/step-role.ts}: guess a step's role from its neighbours in
 * document order.
 */
public final class StepRole {

    private StepRole() {}

    /**
     * The kinds of the steps immediately {@code before} and {@code after} the step being
     * inferred, in document order.
     */
    public record Neighbours(List<StepKind> before, List<StepKind> after) {
        public Neighbours {
            before = List.copyOf(before);
            after = List.copyOf(after);
        }
    }

    /**
     * Guesses a step's role from its neighbours, using the canonical document order
     * context &rarr; action &rarr; sensor. Purely structural — never inspects sentence
     * words (no Given/When/Then heuristics). The generated snippet always offers the
     * other roles as commented alternatives, so a wrong guess is cheap to correct.
     */
    public static StepKind inferStepRole(Neighbours neighbours) {
        List<StepKind> before = neighbours.before();
        List<StepKind> after = neighbours.after();
        if (after.isEmpty()) {
            return StepKind.SENSOR;
        }
        if (after.contains(StepKind.SENSOR)
                && !before.contains(StepKind.ACTION)
                && !after.contains(StepKind.ACTION)) {
            return StepKind.ACTION;
        }
        if (before.isEmpty()) {
            return StepKind.CONTEXT;
        }
        return StepKind.ACTION;
    }
}
